blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,784 @@
+If this gets changed in Julia, then all the ones marked "ambiguous" should magically disappear.
+Consequently, we can turn our attention to other cases.
+
+Let's look at the next item up the list:
+
+```julia-repl
+julia> tree = trees[end-1]
+insert reduce_empty(::typeof(Base.mul_prod), ::Type{F}) where F<:FixedPoint in FixedPointNumbers at /home/tim/.julia/packages/FixedPointNumbers/w2pxG/src/FixedPointNumbers.jl:225 invalidated:
+   backedges: MethodInstance for reduce_empty(::Function, ::Type{T} where T) triggered MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber}) (136 children) more specific
+```
+
+`reduce_empty(::typeof(Base.mul_prod), ::Type{F}) where F<:FixedPoint` is strictly more specific than `reduce_empty(::Function, ::Type{T} where T)`.
+This might look like one of those "necessary" invalidations.
+However, even though it's marked "more specific," the new method `reduce_empty(::typeof(Base.add_sum), ::Type{F}) where F<:FixedPoint` can't be reached from a call `reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber})`:
+
+```julia-repl
+julia> mi, node = tree[:backedges, 1]
+MethodInstance for reduce_empty(::Function, ::Type{T} where T) => MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber}) at depth 0 with 143 children
+
+julia> node.mi      # this is the MethodInstance that called `mi`
+MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber})
+
+julia> typeintersect(tree.method.sig, node.mi.specTypes)
+Union{}
+```
+
+This is is a consequence of the fact that Julia's compiler has chosen not to specialize the argument types, and `reduce_empty(::Function, ::Type{T} where T)` is broader than what could be determined from the caller.
+Thus, while it looks like a case of greater specificity, in fact this is more analogous to the "partial specialization" described below.
+
+Moving backward another step, we get to `sizeof(::Type{X}) where X<:FixedPoint`.
+Simply put, this looks like a method that we don't need; perhaps it dates from some confusion, or an era where perhaps it was necessary.
+So we've discovered an easy place where a developer could do something to productively decrease the number of invalidations, in this case by just deleting the method.
+
+You'll also notice one example where the new method is *less specific*.
+It is not clear why such methods should be invalidating, and this may be a Julia bug.
+
+### Partial specialization
+
+If you try
+
+```julia-repl
+julia> trees = invalidation_trees(@snoopr using StaticArrays)
+```
+
+you'll see a much longer output.
+A large number of invalidations derive from the fact that StaticArrays defines a method for `!=`, which invalidates the fallback definition
+
+```
+!=(x, y) = !(x == y)
+```
+
+Since such definitions account for hundreds of nominal invalidations, it would be well worth considering whether it is possible to delete the custom `!=` methods.
```

Comment:
Fascinating!  And the reason that `==(::Any, ::Any)` doesn't have any dependencies is surely because its method table is so huge that doesn't do any call-site optimizations, yeah?

My initial thoughts were that defining `==` would just cascade its invalidations through `!=`... but that's not true because there's not a backedge for `(::Any, ::Any)` there!  And that's because the call-site optimization happened on `!=(::Any,::Any)` and the `==` is safely generically called within that method.